### PR TITLE
Warn When The HTML Mismatches in DEV

### PR DESCRIPTION
**Pull request by sebmarkbage**
In the new hydration strategy we do our best to recover when the server response isn't valid but we don't check the attributes. However, we should still warn in DEV because that can end up pretty bad.

This starts adding warnings for those cases. I still have several steps left but this can ideally be landed at any midway point if someone wants to review it.

- [x] Warn for differences in text content.
- [x] Warn for known props that have different values.
- [x] Warn for extra attributes that are not known.
- [x] Warn for dangerouslySetInnerHTML differences.
- [x] Warn for extra nodes that will be deleted.
- [x] Warn for missing nodes that will have a new node inserted.

Left this for a follow up PR.

- [ ] Warn for inline styles that have different values.
- [ ] Add more text coverage for individual attribute mismatches, insert/delete, style mismatches, single child text content mismatches.
- [ ] Consolidate multiple warnings into a single warn (ideally with a HTML diff like format) instead of just warning for one difference at a time.

### CODE REVIEW

File: scripts/fiber/tests-passing-except-dev.txt [hunk 1]
```diff
@@ -5,150 +5,18 @@ src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js
...
-* should error reconnecting missing children
-* should error reconnecting added children
-* should error reconnecting more children
-* should error reconnecting fewer children
-* should error reconnecting reordered children
-* should error reconnecting a div with children separated by whitespace on the client
-* should error reconnecting a div with children separated by different whitespace on the server
-* should error reconnecting a div with children separated by different whitespace
-* can distinguish an empty component from a dom node
+* renders a textarea with a value and an onChange with client render on top of good server markup
```
Comment (sebmarkbage) [inline scripts/fiber/tests-passing-except-dev.txt]: We currently apparently removing the text node from the textarea even though we really just manage the `value` property. Need to fix that by not trying to hydrate textarea content. Might be related to dangerouslySetInnerHTML.

File: scripts/fiber/tests-passing-except-dev.txt [hunk 2]
```diff
@@ -5,150 +5,18 @@ src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js
...
+* renders a textarea with a value and readOnly with client render on top of good server markup
+* renders a textarea with a value and no onChange/readOnly with client render on top of good server markup
+* renders a textarea with a defaultValue with client render on top of good server markup
+* renders a textarea value overriding defaultValue with client render on top of good server markup
+* renders a textarea value overriding defaultValue no matter the prop order with client render on top of good server markup
+* renders a controlled textarea with client render on top of good server markup
+* throws when rendering if getChildContext exists without childContextTypes with client render on top of bad server markup
+* throws when rendering if getChildContext returns a value not in childContextTypes with client render on top of bad server markup
```
Comment (sebmarkbage) [inline scripts/fiber/tests-passing-except-dev.txt]: These might need to be updated. This is issuing a warning because the rendered content of an error boundary is not the same content as the server rendered content.

**Comment by sebmarkbage** [comment]
@aickin This basically would've been a broken mess without your tests. So super helpful to find edge cases.

File: scripts/fiber/tests-passing-except-dev.txt [hunk 3]
```diff
@@ -5,150 +5,18 @@ src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js
 * should suggest property name if available (ssr)
 
 src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js
-* renders a blank div with client render on top of bad server markup
-* renders a div with inline styles with client render on top of bad server markup
-* renders a self-closing tag with client render on top of bad server markup
-* renders a self-closing tag as a child with client render on top of bad server markup
-* renders simple numbers with client render on top of bad server markup
-* renders simple strings with client render on top of bad server markup
-* renders string prop with true value with client render on top of bad server markup
```
Comment (sebmarkbage) [inline scripts/fiber/tests-passing-except-dev.txt]: I guess all of these should test these particular values can issue warnings, but in reality they all just test that the tag name didn't match. Which is why this fixes all of them.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 4]
```diff
@@ -69,6 +70,49 @@ if (__DEV__) {
...
+      return;
+    }
+    didWarnInvalidHydration = true;
+    warning(
+      false,
+      'Text content did not match. Server: "%s" Client: "%s"',
+      serverText,
+      clientText,
+    );
+  };
```
Comment (nhunzaker) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Sort of beside the point, but I'm curious if this warning gets lifted out with uglify. Has there been any consideration to a `warnOnce`?

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 5]
```diff
@@ -766,6 +810,36 @@ var ReactDOMFiberComponent = {
...
+          case 'data-react-checksum':
+            break;
+          // Controlled attributes are not validated
+          // TODO: Only ignore them on controlled tags.
+          case 'value':
+            break;
+          case 'checked':
+            break;
+          case 'selected':
+            break;
```
Comment (nhunzaker) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Could these be collapsed into a single `break;`, like:

```
case 'data-reactroot':
case 'data-reactid':
  // etc
  break;
```
?

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js (hunk 4 above)
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Even if hoisted, this assignment would get stripped, so it'd still get DCE.

We have had a few warnOnce solutions but the context of "once" varies a lot so hasn't been worth the abstraction cost. A few:

- Once per React module instance.
- Once per component class.
- Once per component instance.
- Once per subtree.

These can then combine with:

- Once per each type of warning.
- Once for any of multiple types of warnings.
- Etc.

In this case, we should ideally gather all of these warnings up into one actionable "diff view" of all the differences. That's on my todo. To avoid spam, right now I just want to show the first diff among any type of diff.

Comment (nhunzaker) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Awesome, that's great to know. Very cool.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js (hunk 5 above)
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: We have a lint rule against fall through.

Comment (acdlite) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: We fall through elsewhere https://github.com/facebook/react/blob/master/src/renderers/shared/fiber/ReactFiberBeginWork.js#L864-L869

Comment (nhunzaker) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: I was also drawing from:
https://github.com/facebook/react/blob/master/src/renderers/dom/shared/eventPlugins/SimpleEventPlugin.js#L156-L186

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 6]
```diff
@@ -69,6 +70,66 @@ if (__DEV__) {
...
+
+  var warnForExtraAttributes = function(attributeNames: Set<string>) {
+    if (didWarnInvalidHydration) {
+      return;
+    }
+    didWarnInvalidHydration = true;
+    var names = [];
+    attributeNames.forEach(function(name) {
+      names.push(name);
+    });
```
Comment (acdlite) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Interesting, I wouldn't have guessed that `forEach` was the fastest way to do this. TIL. https://jsperf.com/set-iterator-vs-foreach/4

Comment (marvinhagemeister) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Just ran the benchmarks from that link locally in node `8.1.3` and the iterator wins there by a tiny bit. I feel like microbenchmarks cannot be trusted in js. Perhaps @bmeurer can chime in?

Comment (bmeurer) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: Unfortunately both are pretty slow in V8 at the moment. The fact that `Set.prototype.forEach` is faster in some cases isn't intended, and I wouldn't rely on that. We plan to improve the performance of the collection iterators in the future BTW.

My suggestion here: Don't base decisions on current perf results. Use whatever makes sense for your code and try to improve readability instead.

Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: This isn't a perf hot path at all. It only exist in development mode and only fires if something already went wrong.
